cpp/src_prims/sparse/distance/operators.cuh

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/device_atomics.cuh>
+
+namespace MLCommon {
+namespace Sparse {
+namespace Distance {
+
+struct Sum {
+  template <typename value_t>
+  __host__ __device__ __forceinline__ value_t operator()(value_t a, value_t b) {
+    return a + b;
+  }
+};
+
+struct SqDiff {
+  template <typename value_t>
+  __host__ __device__ __forceinline__ value_t operator()(value_t a, value_t b) {
+    return (a - b) * (a - b);
+  }
+};
+
+struct PDiff {
+  float p;
+
+  PDiff(float p_) : p(p_) {}
+
+  template <typename value_t>
+  __host__ __device__ __forceinline__ value_t operator()(value_t a, value_t b) {
+    return powf(a - b, p);
+  }
+};
+
+struct Max {
+  template <typename value_t>
+  __host__ __device__ __forceinline__ value_t operator()(value_t a, value_t b) {
+    return fmaxf(a, b);
+  }
+};
+
+struct AtomicAdd {
+  template <typename value_t>
+  __host__ __device__ __forceinline__ value_t operator()(value_t *a,
```

Comment:
Can this and other ones which use CUDA primitives be valid with the `__host__` identifier? If not, it would be nice to just use `__device__` to be explicit about it

Reply:
Cub uses `__host__ __device__ __forceinline__` in their functors and I don't see much benefit in removing these. The benefit to keeping them is that we don't get any unexpected surprises when it needs to be called from host code, such as when debugging or testing.